### ISSUE TITLE
Add a TaggedUnion runtype.

### ISSUE THREAD

**Issue by mscharley** [pelotom/runtypes#67]
There's a fairly common pattern for using union types called tagged unions. Using the current `Union` runtype for these is functionally complete, however when a `.check()` fails, it's usually near impossible to work out why. I'd like to propose a new `TaggedUnion` runtype. You could use it as follows:

```typescript
const AB = TaggedUnion("tag", {
  a: Record({ aprop: String }),
  b: Record({ bprop: Number }),
});

AB.check({ tag: "a", aprop: "foo" }); // success.
AB.check({ tag: "b", bprop: 10 }); // success.
AB.check({ tag: "c", cprop: "bar" }); // Error: Unknown tag: Expected "a" | "b", but was "c"
AB.check({ tag: "a", bprop: 10 }); // Error: Expected { tag: "a"; aprop: string; }, but was object.

AB.checkTag({ tag: "a" }); // true
AB.check({ tag: "a" }); // Error: Expected { tag: "a"; aprop: string; }, but was object.
AB.checkTag({ tag: "c", cprop: "bar" }); // false
```

Having a seperate `checkTag` is often useful for deserialising input from changing upstream providers such as webhooks where they may add new event types which you want to discard or log differently because you typically care less about them, but you do care if you get an event you thought you knew about which has a structure other than you were expecting.

As an example of why I feel this is necessary, the current approach leaves me with errors like the following while trying to deserialise Mailgun webhooks (if there is an error):

> Error: Expected (((({ campaigns: always[]; id: string; log-level: string; message: { headers: { message-id: string; } & { from?: string; subject?: string; to?: string; }; }; recipient: string; tags: string[]; timestamp: number; user-variables: { cortex?: string; }; } & { recipient-domain?: string; }) & { envelope: { sender?: string; sending-ip?: string; targets?: string; transport?: string; }; flags: { is-authenticated?: boolean; is-routed?: boolean; is-system-test?: boolean; is-test-mode?: boolean; }; message: { attachments: {}[]; size: number; }; }) & { delivery-status: { attempt-no: number; code: number; description: string; message: string; session-seconds: number; } & { certificate-verified?: boolean; mx-host?: string; retry-seconds?: number; tls?: boolean; utf8?: boolean; }; storage: { key: string; url: string; }; }) & { event: "delivered"; }) | ((({ campaigns: always[]; id: string; log-level: string; message: { headers: { message-id: string; } & { from?: string; subject?: string; to?: string; }; }; recipient: string; tags: string[]; timestamp: number; user-variables: { cortex?:string; }; } & { recipient-domain?: string; }) & { client-info: { client-name: string; client-os: string; client-type: string; device-type: string; user-agent: string; }; geolocation: { city: string; country: string; region: string; }; ip: string; }) & { event: "opened"; }) | ((({ campaigns: always[]; id: string; log-level: string; message: { headers: { message-id: string; } & { from?: string; subject?: string; to?: string; }; }; recipient: string; tags: string[]; timestamp: number; user-variables: { cortex?: string; }; } & { recipient-domain?: string; }) & { client-info: { client-name: string; client-os: string; client-type: string; device-type: string; user-agent: string; }; geolocation: { city: string; country: string; region: string; }; ip: string; }) & { event: "clicked"; url: string; }) | ((({ campaigns: always[]; id: string; log-level: string; message: { headers: { message-id: string; } & { from?: string; subject?: string; to?: string; }; }; recipient: string; tags: string[]; timestamp: number; user-variables: { cortex?: string; }; } & { recipient-domain?: string; }) & { envelope: { sender?: string; sending-ip?: string; targets?: string; transport?: string; }; flags: { is-authenticated?: boolean; is-routed?: boolean; is-system-test?: boolean; is-test-mode?: boolean; }; message: { attachments: {}[]; size: number; }; }) & { event: "complained"; }) | ((({ campaigns: always[]; id: string; log-level: string; message: { headers: { message-id: string; } & { from?: string; subject?: string; to?: string; }; }; recipient: string; tags: string[]; timestamp: number; user-variables: { cortex?: string; }; } & { recipient-domain?: string; }) & { client-info: { client-name: string; client-os: string; client-type: string; device-type: string; user-agent: string; }; geolocation: { city: string; country: string; region: string; }; ip: string; }) & { event: "unsubscribed"; }) | (((({ campaigns: always[]; id: string; log-level: string; message: { headers: { message-id: string; } & { from?: string; subject?: string; to?: string; }; }; recipient: string; tags: string[]; timestamp: number; user-variables: { cortex?: string; }; } & { recipient-domain?: string; }) & { envelope: { sender?: string; sending-ip?: string; targets?: string; transport?: string; }; flags: { is-authenticated?: boolean; is-routed?: boolean; is-system-test?: boolean; is-test-mode?: boolean; }; message: { attachments: {}[]; size: number; }; }) & { delivery-status: { attempt-no: number; code: number; description: string; message: string; session-seconds: number; } & { certificate-verified?: boolean; mx-host?: string; retry-seconds?: number; tls?: boolean; utf8?: boolean; }; storage: { key: string; url: string; }; }) & { event: "failed"; reason: string; severity: "temporary"; }) | (((({ campaigns: always[]; id: string; log-level: string; message: { headers: { message-id: string; } & { from?: string; subject?: string; to?: string; }; }; recipient: string; tags: string[]; timestamp: number; user-variables: { cortex?: string; }; } & { recipient-domain?: string; }) & { envelope: { sender?: string; sending-ip?: string; targets?: string; transport?: string; }; flags: { is-authenticated?: boolean; is-routed?: boolean; is-system-test?: boolean; is-test-mode?: boolean; }; message: { attachments: {}[]; size: number; }; }) & { delivery-status: { attempt-no: number; code: number; description: string; message: string; session-seconds: number; } & { certificate-verified?: boolean; mx-host?: string; retry-seconds?: number; tls?: boolean; utf8?: boolean; }; storage: { key: string; url: string; }; }) & { event: "failed"; reason: string; severity: "permanent"; }), but was object

**Comment by pelotom** [comment]
Yeah, this has been on my TODO list for a while... there are also lots of nice operators you can autogenerate when you know you're working with tagged unions. I'm thinking more and more than integration with [unionize](https://github.com/pelotom/unionize) might be well worth it.

**Comment by mscharley** [comment]
Hmm, unionise looks interesting. But it looks like unionize already has type guards baked in, what sort of integration were you considering?

**Comment by pelotom** [comment]
Unionize has a more succinct and efficient `match` construct for the special case when you're dealing with tagged unions.

**Comment by mscharley** [comment]
I suppose the question was more: What sort of integration would you do between `unionize` and `runtypes`? Or would you just recommend using unionize over runtypes if you need tagged unions?

**Comment by pelotom** [comment]
Just being able to generate a unionized object from a `TaggedUnion` construct in Runtypes.

**Comment by yuhr** [comment]
Seems already resolved by #149, closing.
